Add allow empty translations config to translate behavior

AD7six opened the pull request:
This reverts the behavior changes (but not the tests) introduced in #5822, to address #5795.

I've defaulted the config value to `true` - as that's how it currently works. I think the default is effectively an arbitrary choice, as it's equally expected/valid to want or not want empty translations to be retrieved.

AD7six committed:
Revert "Merge pull request #5822 from patrickconroy/3.0-translate-fieldconditions"

This only reverts changes made to the behavior, keeping changes made to
tests.

This reverts commit e8d88a7efadafc16414e50b573530075df48bc07, reversing
changes made to 902757ff6b0bc5d20815fd66f73785e7a7fe2568.

Conflicts:
	src/ORM/Behavior/TranslateBehavior.php

AD7six committed:
Adapt to fit expected implementation

AD7six committed:
Add allowEmptyTranslations

This re-implements conditions.

markstory commented:
Looks good to me.

jadb commented on src/ORM/Behavior/TranslateBehavior.php:
```diff
@@ -158,6 +143,14 @@ public function setupFieldAssociations($fields, $table, $fieldConditions, $strat
+                $name . '.model' => $model,
+                $name . '.field' => $field,
+            ];
+            if (!$this->_config['allowEmptyTranslations']) {
```
Why not use `$this->config()`?

AD7six replied:
There are (almost) no other uses of `config()` in the behavior - just following suit. Also, there's no possibility of these keys being missing.

patrickconroy commented on src/ORM/Behavior/TranslateBehavior.php:
```diff
@@ -167,11 +160,16 @@ public function setupFieldAssociations($fields, $table, $fieldConditions, $strat
+        $conditions = ["$targetAlias.model" => $model];
+        if (!$this->_config['allowEmptyTranslations']) {
+            $conditions["$targetAlias.content !="] = '';
+        }
```
This needs to be 
```
if (!$this->_config['allowEmptyTranslations']) {
	$conditions["$targetAlias.content <>"] = '';
}
```
Since that'll check for empty or null.

AD7six replied:
That sounds like misinformation to me =). Which database does not treat `<>` and `!=` as equivalent?

Mysql 5.6 does for example:

```
mysql> create table example ( title varchar(10), thing varchar(10) default null );
Query OK, 0 rows affected (0.02 sec)

mysql> insert into example values('null', null), ('empty', ''), ('stuff', 'stuff');
Query OK, 3 rows affected (0.00 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> select * from example where thing <> '';
+-------+-------+
| title | thing |
+-------+-------+
| stuff | stuff |
+-------+-------+
1 row in set (0.00 sec)

mysql> select * from example where thing != '';
+-------+-------+
| title | thing |
+-------+-------+
| stuff | stuff |
+-------+-------+
1 row in set (0.00 sec)

mysql>
```

in any event, null translation records are ignored [by the row mapper](https://github.com/cakephp/cakephp/blob/3.0-translate-allow-empty/src/ORM/Behavior/TranslateBehavior.php#L389)

patrickconroy replied:
Sorry it's confusingly necessary for the PR I put in for this. I needed the field to not be returned to work nicely when using `filterByCurrentLocale`. Without the bit operation this would happen:
1. Table translated, `onlyTranslated` false
2. `Table->find('translations', ['filterByCurrentLocale' => true]);`
3. This returns records with empty non-null `content`.
4. Those records can't come back since they're not fully translated BUT the behavior can't be set with `onlyTranslated` true, since I need that to be false when working on a backend of possibly untranslated records.

AD7six replied:
I don't understand the above.

>  Without the bit operation

What bit operation?

patrickconroy replied:
Whoops sorry not bit operation, the `<>` check.

AD7six replied:
The not equal operator is either syntax, unless you're using [access or db2](http://stackoverflow.com/a/723426) in which case - you probably have 98 more problems.

Would you mind demonstrating "a difference", pointing at a reference?

patrickconroy replied:
Re-reading I meant to illustrate the difference between these two:
```
select * from foo where name = '';
select * from foo where name IS NULL;
```
Negating an empty string (`content!=''`) returns non-null/non-empty but when checking for an empty string (`content=''`) it'll return non-null/empty. I've been trying to get a query to break without `<>` but `!=''` seems to be doing the right thing. If I can pinpoint where/why `<>` was needed I'll bring it up again. http://stackoverflow.com/questions/14575807/how-to-select-non-empty-columns-in-mysql is a reference to null/empty checking with `<>`.

patrickconroy commented on src/ORM/Behavior/TranslateBehavior.php:
```diff
@@ -473,14 +471,14 @@ protected function _bundleTranslatedFields($entity)
         $results = $this->_findExistingTranslations($find);
-        $model = $this->config('conditions.model');
+        $alias = $this->_table->alias();
 
```
If this is just the table alias, `$this->config('model')` will be useless, it needs to be the model set in the config, and if no model is set it needs to then be the table alias. The way you've changed it means if I've added this behavior, from say `PostsTable` and set the `model` in the config to `Articles`, this wouldn't find any existing translations since in the DB the `model` is `Articles`.

AD7six replied:
I might be misunderstanding, that sounds like the model key is a hack to account for something we're missing. 

Are you saying that this won't work:

    $table = TableRegistry::get('Posts');
    $table->addBehavior('Translate', ...);
    $table->alias('AnythingAtAll');
    $untranslatedRecords = $table->find('all');

?

In 2.x the translate behavior uses the [model name](https://github.com/cakephp/cakephp/blob/master/lib/Cake/Model/Behavior/TranslateBehavior.php#L245) - rather than the model alias.

@markstory @lorenzo should we have a name getter for table objects? That's what structural behaviors would read, rather than the mutable alias. Right now it's [buried in the alias method](https://github.com/cakephp/cakephp/blob/3.0/src/ORM/Table.php#L354-L356)

In any event we have no relevant tests for the model key - which is in part why I was able to know nothing about it and have no tests fail.

AD7six replied:
OR why don't we use `$table->table()` - that's after all the fundamental "thing" that's being translated.

patrickconroy replied:
I'm saying that when `config['model']` was added (https://github.com/cakephp/cakephp/commit/2f16e0c8) it was meant to be the actual value of `model` in the DB for the translated records. Someone needed it to be customizable instead of using the table alias.

AD7six replied:
I've propsed `Table::name()` in #5984 which would, I think, remove the need for the model key.

AD7six committed:
Use the model key if set

AD7six commented:
I've changed the missed use of the `model` key - to maintain the existing behavior but [this discussion](https://github.com/cakephp/cakephp/pull/5983#discussion_r25560677) is still relevant/pending.

AD7six committed:
Another missed 'model' use

AD7six committed:
Consolidate model/alias shuffling

So that if we can make it cleaner - we do so in one place only

AD7six committed:
Rename model to referenceName

While 'model' is the field name, as a config key it is quite meaningless
without looking how it's used. Changed to reference name as, while not
perfect, it's more descriptive than "model".

AD7six committed:
Accurately derive the reference name

So that only in none-conventional/exceptional circumstances is it
necessary for developers to specify it at all

AD7six commented:
After discussing with Jose on irc, I've [added to this pr](https://github.com/cakephp/cakephp/compare/62cfe3c97ddee77849a85160d12665304b3f0a9f...a764fcffe7a960b686c7beaa05e35f4e77cf0e50):

 * renaming the 'model' key to `referenceName` (_slightly_ more descriptive)
 * changed the behavior such that the default `referenceName` still applies whatever the alias/registryAlias a table object has.

That aught to mean that users don't need to change/define `referenceName` at all in normal use.

AD7six commented on src/ORM/Behavior/TranslateBehavior.php:
```diff
@@ -74,10 +75,10 @@ class TranslateBehavior extends Behavior
-        'model' => '',
+        'referenceName' => '',
+        'allowEmptyTranslations' => true,
         'onlyTranslated' => false,
```
Should this be named `returnEmptyTranslations` and `returnOnlyTranslated`(or similar)?

lorenzo replied:
I think the current name is fine

patrickconroy commented:
Yeah, `referenceName` makes sense to me.

lorenzo commented:
Good job @AD7six !
